runtime/parachains/src/scheduler.rs

```diff
@@ -487,77 +517,106 @@ impl<T: Config> Pallet<T> {
 			let cq = ClaimQueue::<T>::get();
 
 			for core_idx in 0..n_session_cores {
-				let core_idx = CoreIndex(core_idx);
-				let group_idx = Self::group_assigned_to_core(core_idx, now).expect(
-					"core is not out of bounds and we are guaranteed \
-										  to be after the most recent session start; qed",
-				);
+				let core_idx = CoreIndex::from(core_idx);
 
 				// add previously timedout paras back into the queue
-				if let Some(assignment) = timedout_paras.remove(&core_idx) {
-					let ca = assignment.to_core_assignment(core_idx, group_idx);
-					Self::add_to_claimqueue(ca);
+				if let Some(mut entry) = timedout_paras.remove(&core_idx) {
+					if entry.retries < T::AssignmentProvider::get_max_retries(core_idx) {
+						entry.retries += 1;
+						Self::add_to_claimqueue(core_idx, entry);
+					} else {
+						// Consider max retried parathreads as concluded for the assignment provider
+						let ret = concluded_paras.insert(core_idx, entry.para_id());
+						debug_assert!(ret.is_none());
+					}
 				}
 
-				let n_lookahead_used = cq.get(&core_idx).map_or(0, |v| v.len() as u32);
+				// We  consider occupied cores to be part of the claimqueue
+				let n_lookahead_used = cq.get(&core_idx).map_or(0, |v| v.len() as u32) +
+					if Self::is_core_occupied(core_idx) { 1 } else { 0 };
 				for _ in n_lookahead_used..n_lookahead {
 					let concluded_para = concluded_paras.remove(&core_idx);
-					if let Some(ass) =
+					if let Some(assignment) =
 						T::AssignmentProvider::pop_assignment_for_core(core_idx, concluded_para)
 					{
-						let ca = ass.to_core_assignment(core_idx, group_idx);
-						Self::add_to_claimqueue(ca);
+						let ttl = <configuration::Pallet<T>>::config().on_demand_ttl;
```

Comment:
Oh oh. This looks off. We are querying the ttl value for on_demand, but the core could be occupied by anything. In the scheduler we should not know anything about on-demand.

Reply:
If there is no other way, this needs to be a function of the `AssignmentProvider` trait. If we go that route, maybe we should combine all those functions into a `core_config()` function or something.